File: src/weaponsprocurement/core/ObservedStockIndex.java
package weaponsprocurement.core;

import com.fs.starfarer.api.campaign.SectorAPI;
import com.fs.starfarer.api.campaign.econ.EconomyAPI;
import com.fs.starfarer.api.campaign.econ.MarketAPI;
import com.fs.starfarer.api.impl.campaign.ids.Submarkets;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class ObservedStockIndex {
    private final MarketStockService marketStockService = new MarketStockService();

    public Map<String, ObservedItem> collect(SectorAPI sector, WeaponMarketBlacklist blacklist) {
        EconomyAPI economy = sector == null ? null : sector.getEconomy();
        List<MarketAPI> markets = economy == null ? null : economy.getMarketsCopy();
        if (markets == null || markets.isEmpty()) {
            return Collections.emptyMap();
        }

        Map<String, ObservedItem> result = new HashMap<String, ObservedItem>();
        for (int i = 0; i < markets.size(); i++) {
            MarketAPI market = markets.get(i);
            MarketStockService.MarketStock stock = marketStockService.collectCurrentMarketItemStock(market, true);
            for (String itemKey : stock.itemKeys()) {
                if (!isEligible(itemKey, blacklist)) {
                    continue;
                }
                List<SubmarketWeaponStock> sources = stock.getSubmarketStocks(itemKey);
                for (int j = 0; j < sources.size(); j++) {
                    SubmarketWeaponStock source = sources.get(j);
                    if (source == null || source.getCount() <= 0 || !source.isPurchasable()) {
                        continue;
                    }
                    ObservedItem current = result.get(itemKey);
                    boolean vanillaSupported = isVanillaSupportedSubmarket(source.getSubmarketId());
                    if (current == null) {
                        result.put(itemKey, new ObservedItem(source, vanillaSupported));
                    } else {
                        current.observe(source, vanillaSupported);
                    }
                }
            }
        }
        return result;
    }

    private static boolean isEligible(String itemKey, WeaponMarketBlacklist blacklist) {
        return FixerMarketObservedCatalog.isSafeFixerItem(itemKey)
                && (blacklist == null || !blacklist.isBannedFromFixers(itemKey));
    }

    static boolean isVanillaSupportedSubmarket(String submarketId) {
        return Submarkets.SUBMARKET_OPEN.equals(submarketId)
                || Submarkets.GENERIC_MILITARY.equals(submarketId)
                || Submarkets.SUBMARKET_BLACK.equals(submarketId);
    }

    public static final class ObservedItem {
        private SubmarketWeaponStock cheapest;
        private boolean vanillaSupportedSeen;
        private boolean unsupportedSeen;

        private ObservedItem(SubmarketWeaponStock source, boolean vanillaSupported) {
            observe(source, vanillaSupported);
        }

        private void observe(SubmarketWeaponStock source, boolean vanillaSupported) {
            if (vanillaSupported) {
                vanillaSupportedSeen = true;
            } else {
                unsupportedSeen = true;
            }
            if (cheapest == null || compareReferenceSource(source, cheapest) < 0) {
                cheapest = source;
            }
        }

        public SubmarketWeaponStock getCheapestReferenceSource() {
            return cheapest;
        }

        public boolean isOnlyUnsupportedCustomSubmarket() {
            return unsupportedSeen && !vanillaSupportedSeen;
        }

        private static int compareReferenceSource(SubmarketWeaponStock left, SubmarketWeaponStock right) {
            int result = Integer.compare(left.getBaseUnitPrice(), right.getBaseUnitPrice());
            if (result != 0) {
                return result;
            }
            return left.getDisplaySourceName().compareToIgnoreCase(right.getDisplaySourceName());
        }
    }
}

File: src/main/kotlin/weaponsprocurement/core/ObservedStockIndex.kt
package weaponsprocurement.core

import com.fs.starfarer.api.campaign.SectorAPI
import com.fs.starfarer.api.impl.campaign.ids.Submarkets
import java.util.HashMap

class ObservedStockIndex {
    private val marketStockService = MarketStockService()

    fun collect(sector: SectorAPI?, blacklist: WeaponMarketBlacklist?): Map<String, ObservedItem> {
        val economy = sector?.economy
        val markets = economy?.marketsCopy
        if (markets == null || markets.isEmpty()) return emptyMap()

        val result = HashMap<String, ObservedItem>()
        for (market in markets) {
            val stock = marketStockService.collectCurrentMarketItemStock(market, true)
            for (itemKey in stock.itemKeys()) {
                if (!isEligible(itemKey, blacklist)) continue
                val sources = stock.getSubmarketStocks(itemKey)
                for (source in sources) {
                    if (source == null || source.count <= 0 || !source.isPurchasable()) continue
                    val current = result[itemKey]
                    val vanillaSupported = isVanillaSupportedSubmarket(source.submarketId)
                    if (current == null) {
                        result[itemKey] = ObservedItem(source, vanillaSupported)
                    } else {
                        current.observe(source, vanillaSupported)
                    }
                }
            }
        }
        return result
    }

    class ObservedItem private constructor(
        private var cheapest: SubmarketWeaponStock?,
        private var vanillaSupportedSeen: Boolean,
        private var unsupportedSeen: Boolean,
    ) {
        constructor(source: SubmarketWeaponStock, vanillaSupported: Boolean) : this(null, false, false) {
            observe(source, vanillaSupported)
        }

        fun observe(source: SubmarketWeaponStock, vanillaSupported: Boolean) {
            if (vanillaSupported) {
                vanillaSupportedSeen = true
            } else {
                unsupportedSeen = true
            }
            val currentCheapest = cheapest
            if (currentCheapest == null || compareReferenceSource(source, currentCheapest) < 0) {
                cheapest = source
            }
        }

        val cheapestReferenceSource: SubmarketWeaponStock?
            get() = cheapest

        val isOnlyUnsupportedCustomSubmarket: Boolean
            get() = unsupportedSeen && !vanillaSupportedSeen

        companion object {
            private fun compareReferenceSource(left: SubmarketWeaponStock, right: SubmarketWeaponStock): Int {
                val result = left.baseUnitPrice.compareTo(right.baseUnitPrice)
                if (result != 0) return result
                return (left.displaySourceName ?: "").compareTo(right.displaySourceName ?: "", ignoreCase = true)
            }
        }
    }

    companion object {
        private fun isEligible(itemKey: String, blacklist: WeaponMarketBlacklist?): Boolean {
            return FixerMarketObservedCatalog.isSafeFixerItem(itemKey) &&
                (blacklist == null || !blacklist.isBannedFromFixers(itemKey))
        }

        @JvmStatic
        fun isVanillaSupportedSubmarket(submarketId: String?): Boolean {
            return Submarkets.SUBMARKET_OPEN == submarketId ||
                Submarkets.GENERIC_MILITARY == submarketId ||
                Submarkets.SUBMARKET_BLACK == submarketId
        }
    }
}
